### PR TITLE
feat: relay context update

### PR DESCRIPTION
**Pull request by 0xiamflux**
Closes OPT-589
Closes OPT-590

### CODE REVIEW

File: packages/contracts-bedrock/src/L2/SuperchainTokenBridge.sol [hunk 1]
```diff
@@ -86,7 +86,7 @@ contract SuperchainTokenBridge {
     function relayERC20(address _token, address _from, address _to, uint256 _amount) external {
         if (msg.sender != MESSENGER) revert Unauthorized();
 
-        (address crossDomainMessageSender, uint256 source) =
+        (address crossDomainMessageSender, uint256 source, /*address entrypoint*/, /*uint256 nonce*/ ) =
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/src/L2/SuperchainTokenBridge.sol]: No need to add those comments
```suggestion
        (address crossDomainMessageSender, uint256 source, ,  ) =
```

File: packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol [hunk 2]
```diff
@@ -210,14 +245,15 @@ contract L2ToL2CrossDomainMessenger is ISemver, TransientReentrancyAware {
...
+            _entrypoint: entrypoint,
             _message: message
         });
 
         if (successfulMessages[messageHash]) {
             revert MessageAlreadyRelayed();
         }
 
-        _storeMessageMetadata(source, sender);
+        _storeMessageMetadata(source, sender, entrypoint, nonce);
```
Comment (0xDiscotech) [inline packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol:256]: To handle the case described here https://discord.com/channels/837716795213217882/1303795552994983936/1312122424820629527:
```suggestion
        _storeMessageMetadata(source, sender, msg.sender, nonce);
```

File: packages/contracts-bedrock/src/L2/SuperchainWETH.sol [hunk 3]
```diff
@@ -150,7 +150,7 @@ contract SuperchainWETH is WETH98, IERC7802, ISemver {
     function relayETH(address _from, address _to, uint256 _amount) external {
         if (msg.sender != Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) revert Unauthorized();
 
-        (address crossDomainMessageSender, uint256 source) =
+        (address crossDomainMessageSender, uint256 source, /*address entrypoint*/, /*uint256 nonce*/ ) =
```
Comment (hexshire) [inline packages/contracts-bedrock/src/L2/SuperchainWETH.sol]: > No need to add those comments

Same here.

**Review by 0xDiscotech — COMMENTED** [review]
Missing `testFuzz_crossDomainMessageNonce` tests for succeed and revert

**Comment by hexshire** [comment]
Some test cases failed
image

File: packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol (hunk 2 above)
Comment (0xDiscotech) [reply packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol:256]: for now, dismiss, we can leave the question to architect.
But adding the sender as entrypoint when it was not set, could mbe confusing or mess up stuff on the target call?

File: packages/contracts-bedrock/src/L2/SuperchainTokenBridge.sol (hunk 1 above)
Comment (0xiamflux) [reply packages/contracts-bedrock/src/L2/SuperchainTokenBridge.sol]: Addressed [here](https://github.com/defi-wonderland/optimism/pull/159/commits/35307d13b4413742580babadb9ea6e9f50ed6e8c).

File: packages/contracts-bedrock/src/L2/SuperchainWETH.sol (hunk 3 above)
Comment (0xiamflux) [reply packages/contracts-bedrock/src/L2/SuperchainWETH.sol]: Addressed [here](https://github.com/defi-wonderland/optimism/pull/159/commits/35307d13b4413742580babadb9ea6e9f50ed6e8c).

**Comment by 0xiamflux** [comment]
> Missing `testFuzz_crossDomainMessageNonce` tests for succeed and revert

Addressed [here](https://github.com/defi-wonderland/optimism/pull/159/commits/2eb5879f8e07bd84519689b8934e532615922973).

**Comment by 0xiamflux** [comment]
> Some test cases failed image

Addressed [here](https://github.com/defi-wonderland/optimism/pull/159/commits/fb0112dd8ee957cae4921b145c0d7d36f39a326d).

**Review by skeletor-spaceman — APPROVED** [review]

**Review by hexshire — APPROVED** [review]

**Review by 0xDiscotech — APPROVED** [review]

**Pull request merged**